src/goose/brokers/redis/batch.clj

```diff
@@ -0,0 +1,191 @@
+  ;; `when-let` guards against nil return from `get-batch`.
+  (when-let [{:keys [ready-queue] :as batch} (get-batch redis-conn id)]
+    ;; If a worker dies after a callback is enqueued,
+    ;; and before a job is deleted from orphan-queue, a batch-job
+    ;; becomes orphan and is retried. Before enqueuing callback,
+    ;; we check if a batch isn't already in terminal state.
+    (when-not (batch/terminal-state? (:status batch))
+      (let [{:keys [batch-hash]} (batch-keys id)
+            callback (batch/new-callback batch id status)]
+        (redis-cmds/atomic
+          redis-conn
+          (car/multi)
+          (car/rpush ready-queue callback) ; Enqueue callback to front of queue.
+          (car/hset batch-hash :status status))))))
+
+(defn- job-source-set
+  [job {:keys [enqueued-set retrying-set]}]
+  (if (job/retried? job)
+    retrying-set
+    enqueued-set))
+
+(defn- job-destination-set
+  ([job batch-keys]
+   (job-destination-set job batch-keys nil))
+  ([job {:keys [retrying-set success-set dead-set]} ex]
+   (if-not ex
+     success-set
+     (if (goose.retry/max-retries-reached? job)
+       dead-set
+       retrying-set))))
+
+;;; Updating a batch & getting status in one transaction will guard against
+;;; n final batch-jobs completing at same time concurrently.
+;;; Only 1 of the batch-jobs will receive a terminal status post execution.
+;;; This operation MUST be executed within a transaction, and not a pipeline.
+;;; Reasoning: https://stackoverflow.com/a/29337358/19427982.
+(defn- update-batch-and-get-status-atomically
+  [redis-conn src dst job-id batch-keys status]
+  (let [{:keys [enqueued-set retrying-set success-set dead-set]} batch-keys
+        [_ atomic-results] (redis-cmds/atomic
+                             redis-conn
+                             (car/multi)
+                             (car/smove src dst job-id)
+                             (car/scard enqueued-set)
+                             (car/scard retrying-set)
+                             (car/scard success-set)
+                             (car/scard dead-set))
+        [_ enqueued retrying success dead] atomic-results]
+    (reset! status (batch/status-from-job-states enqueued retrying success dead))))
+
+(defn- execute-batch
+  [next
+   {:keys [redis-conn] :as opts}
+   {job-id :id batch-id :batch-id :as job}]
+  (let [status (atom nil)
+        batch-keys (batch-keys batch-id)
+        src (job-source-set job batch-keys)]
+    (try
+      (let [response (next opts job)
+            dst (job-destination-set job batch-keys)]
+        (update-batch-and-get-status-atomically redis-conn src dst job-id batch-keys status)
+        response)
+      (catch Exception ex
+        (let [failed-job (goose.retry/set-failed-config job ex)
+              dst (job-destination-set failed-job batch-keys ex)]
+          (update-batch-and-get-status-atomically redis-conn src dst job-id batch-keys status)
+          (throw ex)))
+      (finally
```

Comment:
question: does this get called for the exception? after the exception is thrown again

Reply:
This isn't called for the exception, rather to mark a batch as complete regardless of job success/failure.